### PR TITLE
Onboarding: Invite new users on store level

### LINKED ISSUE #5563: New user onboarding flow
For the new user onboarding flow we assume there is an instance to connect to already. 

## User Types

There are at least three roles/use cases for the flow of onboarding/initial login:

1. Admin
    - Their account got created when setting up the instance.
    - They could either set up merchant accounts (and stores) in advance or simply offer a registration.
    - Related: #5548
2. Merchant
    - Their account might have been created by the admin or they get to the instance for registering.
    - Manage the store and access roles for their employees.
    - They provide employee accounts by setting them up in advance or let their employees register.
3. Employee
    - Roles like cashier, with limited store access.

## Onboarding Approaches

1. Open Instance
    - Users register themselves via **shared register link**
    - Admin can require email confirmation and approval
2. Invitation based
    - Admin or store owner creates basic user account with email address
    - User gets an **email with invite link** request to setup the account (confirm email, set password)
        - "You have been invited to INSTANCE/STORE, setup your account here: LINK/QRCODE"

In both scenarios the admin has the option to deactivate user later on.
 
## Invite links

Store Owners and Employees can be provided with a link/qrcode, which is basically an encoded version of ...

- The instance URL to connect to
- An access token we can use on the instance to initially identify them and their setup state, which can either be:
    - Pre-setup by the supervisor role (Admin -> Merchant, Merchant -> Employee)
    - Registration and assignment to the defined role for new users, no prior setup

This is very similar to the flow of connecting a LNDhub-compatible wallet: A custom link scheme which can be handled by our app plus the connection details. Something like this, which can be shared as a link or QR code: 

```
https://SERVER_URL/invite/USER_ID/TOKEN
```

The app can then transfer this into an API call to `https://SERVER_URL/api/app//USER_ID/TOKEN`, which handles the setup and responds with the account details for the app. 

## TODO
- #5714
  - [x] Separate self-registration and invite cases
  - [x] Add invitation email for users created by the admin (right now there is no email when an account is setup with a password)
  - [x] Add handler action for invite links
- #5719
  - [x] Add user creation/invite on the store level (right now one can only add existing users to stores)
  - [x] Improve default role structure

### Optional and related
- #5718 — can be incorporated if/when merged.
  - [x] Make instance name configurable to allow for better branding
  - [x] Add contact email address for instance (see #4806)

- [ ] Integrate everything
- [ ] Wording adjustments, e.g. in emails

### PR DESCRIPTION
**Pull request by dennisreimann**
Includes the updates from #5714 and can be rebased once that PR is merged.

- Add Store Manager and Employee roles
- Roles: Add default order and improve display
- Store users: Invite non-existing users
- Store users: Allow to change user role in place (before you had to remove and re-add the user)

Closes #5563

## Roles
roles

## Inviting users

invite

invite-accepted

### CODE REVIEW

**Review by dstrukt — CHANGES_REQUESTED** [review]
First pass looks good functionally, few thoughts:

- [ ] Think we should change the `Set Password` CTA copy to `Create Account` imo, once the invitation is accepted.

- I don't love the long invite link, think it could be treated like an address, in that it's copyable and better formatted, but the current iteration works and satisfies all the required functionality. I realize this may deserve a mock to play with a few ideas that clean it up .. so maybe we can address this when we add expiration(s) to the invites? (unless this functionality was nACK'd)

- I do like the look of the `Scope` tags, and love that you differentiated them via colors, which is sufficient for this PR. However I'm wondering if we should be doing more to push the look of certain things, e.g. the Server Admin gets a crown, or the Server level has a server icon / Store has a store icon, etc.. thoughts?

Still trying to break this, but haven't been able to yet, the above are minor nits and considerations! Awesome work!

**Comment by pavlenex** [comment]
Began testing but ran into two problems thus far:

1. Store owner doesn't get any notification or email if an invited successfully registered.
2. Seems like permissions are not working correctly, I've invited user as a manager, but he only had guest access and no ability to view store settings.

There are some things in the logs, but unsure what they're saying and if they're PR related:

```
2024-03-04 21:06:48.493 +05:00 [INF] Root Path: /
2024-03-04 21:06:49.908 +05:00 [INF] Connection to NBXplorer's database successful, dashboard and reporting features activated.
2024-03-04 21:06:49.910 +05:00 [INF] SSH settings detected, testing connection to root@127.0.0.1 on port 21622 ...
2024-03-04 21:06:50.109 +05:00 [INF] Checking if any payment arrived on lightning while the server was offline...
2024-03-04 21:06:50.166 +05:00 [INF] SSH connection succeeded
2024-03-04 21:06:50.418 +05:00 [INF] Processing lightning payments...
2024-03-04 21:06:50.518 +05:00 [INF] Starting listening NBXplorer (BTC)
2024-03-04 21:06:50.533 +05:00 [INF] Start watching invoices
2024-03-04 21:06:50.589 +05:00 [INF] Starting payment request expiration watcher
2024-03-04 21:06:51.000 +05:00 [INF] 0 pending payment requests being checked since last run
2024-03-04 21:06:51.080 +05:00 [INF] BTC: Checking if any pending invoice got paid while offline...
2024-03-04 21:06:51.113 +05:00 [INF] BTC: 0 payments happened while offline
2024-03-04 21:06:51.113 +05:00 [INF] Connected to WebSocket of NBXplorer (BTC)
2024-03-04 21:07:20.487 +05:00 [INF] New admin xxxh@icloud.com registered
2024-03-04 21:09:06.872 +05:00 [ERR] Unexpected exception in TimingPipeFlusher.FlushAsync.
System.IO.IOException: The encryption operation failed, see inner exception.
 ---> System.ComponentModel.Win32Exception (14): Bad address
   --- End of inner exception stack trace ---
   at System.Net.Security.SslStream.WriteSingleChunk[TIOAdapter](ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at System.Net.Security.SslStream.WriteAsyncInternal[TIOAdapter](ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at System.IO.Pipelines.StreamPipeWriter.FlushAsyncInternal(Boolean writeToStream, ReadOnlyMemory`1 data, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.PipeWriterHelpers.TimingPipeFlusher.FlushAsync(MinDataRate minRate, Int64 count, IHttpOutputAborter outputAborter, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2FrameWriter.WriteGoAwayAsync(Int32 lastStreamId, Http2ErrorCode errorCode)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.UpdateConnectionState()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
--- End of stack trace from previous location ---
   at System.Net.Security.SslStream.WriteAsyncInternal[TIOAdapter](ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at System.IO.Pipelines.StreamPipeWriter.FlushAsyncInternal(Boolean writeToStream, ReadOnlyMemory`1 data, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.PipeWriterHelpers.ConcurrentPipeWriter.FlushAsyncAwaited(ValueTask`1 flushTask, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.PipeWriterHelpers.TimingPipeFlusher.TimeFlushAsyncAwaited(ValueTask`1 pipeFlushTask, MinDataRate minRate, IHttpOutputAborter outputAborter, CancellationToken cancellationToken)
2024-03-04 21:09:06.894 +05:00 [ERR] Unhandled exception while processing 0HN1SCKE349O6.
System.IO.IOException: The encryption operation failed, see inner exception.
 ---> System.ComponentModel.Win32Exception (14): Bad address
   --- End of inner exception stack trace ---
   at System.Net.Security.SslStream.WriteSingleChunk[TIOAdapter](ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at System.Net.Security.SslStream.WriteAsyncInternal[TIOAdapter](ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at System.IO.Pipelines.StreamPipeWriter.FlushAsyncInternal(Boolean writeToStream, ReadOnlyMemory`1 data, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at System.IO.Pipelines.StreamPipeWriter.FlushAsyncInternal(Boolean writeToStream, ReadOnlyMemory`1 data, CancellationToken cancellationToken)
   at System.IO.Pipelines.StreamPipeWriter.CompleteAsync(Exception exception)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at System.IO.Pipelines.StreamPipeWriter.CompleteAsync(Exception exception)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.DuplexPipeStreamAdapter`1.DisposeAsync()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.DuplexPipeStreamAdapter`1.DisposeAsync()
   at Microsoft.AspNetCore.Server.Kestrel.Https.Internal.HttpsConnectionMiddleware.OnConnectionAsync(ConnectionContext context)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.HttpConnection.ProcessRequestsAsync[TContext](IHttpApplication`1 httpApplication)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
--- End of stack trace from previous location ---
   at System.Net.Security.SslStream.WriteAsyncInternal[TIOAdapter](ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at System.IO.Pipelines.StreamPipeWriter.FlushAsyncInternal(Boolean writeToStream, ReadOnlyMemory`1 data, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
   at System.IO.Pipelines.StreamPipeWriter.CompleteAsync(Exception exception)
   at System.IO.Pipelines.StreamPipeWriter.CompleteAsync(Exception exception)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.DuplexPipeStreamAdapter`1.DisposeAsync()
   at Microsoft.AspNetCore.Server.Kestrel.Https.Internal.HttpsConnectionMiddleware.OnConnectionAsync(ConnectionContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Https.Internal.HttpsConnectionMiddleware.OnConnectionAsync(ConnectionContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.KestrelConnection`1.ExecuteAsync()
2024-03-04 21:29:37.698 +05:00 [INF] Root Path: /
2024-03-04 21:29:39.003 +05:00 [INF] Connection to NBXplorer's database successful, dashboard and reporting features activated.
2024-03-04 21:29:39.005 +05:00 [INF] SSH settings detected, testing connection to root@127.0.0.1 on port 21622 ...
2024-03-04 21:29:39.122 +05:00 [INF] Checking if any payment arrived on lightning while the server was offline...
2024-03-04 21:29:39.169 +05:00 [INF] SSH connection succeeded
2024-03-04 21:29:39.226 +05:00 [INF] Processing lightning payments...
2024-03-04 21:29:39.256 +05:00 [INF] Starting listening NBXplorer (BTC)
2024-03-04 21:29:39.269 +05:00 [INF] Start watching invoices
2024-03-04 21:29:39.310 +05:00 [INF] Starting payment request expiration watcher
2024-03-04 21:29:39.520 +05:00 [INF] 0 pending payment requests being checked since last run
2024-03-04 21:29:39.761 +05:00 [INF] BTC: Checking if any pending invoice got paid while offline...
2024-03-04 21:29:39.816 +05:00 [INF] BTC: 0 payments happened while offline
2024-03-04 21:29:39.816 +05:00 [INF] Connected to WebSocket of NBXplorer (BTC)
2024-03-04 21:30:16.044 +05:00 [INF] A new user just registered xxx@icloud.com (admin)
2024-03-04 21:39:57.611 +05:00 [INF] Root Path: /
2024-03-04 21:39:58.378 +05:00 [INF] Connection to NBXplorer's database successful, dashboard and reporting features activated.
2024-03-04 21:39:58.380 +05:00 [INF] SSH settings detected, testing connection to root@127.0.0.1 on port 21622 ...
2024-03-04 21:39:58.522 +05:00 [INF] Checking if any payment arrived on lightning while the server was offline...
2024-03-04 21:39:58.616 +05:00 [INF] SSH connection succeeded
2024-03-04 21:39:58.666 +05:00 [INF] Processing lightning payments...
2024-03-04 21:39:58.696 +05:00 [INF] Starting listening NBXplorer (BTC)
2024-03-04 21:39:58.706 +05:00 [INF] Start watching invoices
2024-03-04 21:39:58.733 +05:00 [INF] Starting payment request expiration watcher
2024-03-04 21:39:58.840 +05:00 [INF] 0 pending payment requests being checked since last run
2024-03-04 21:39:59.213 +05:00 [INF] BTC: Checking if any pending invoice got paid while offline...
2024-03-04 21:39:59.233 +05:00 [INF] BTC: 0 payments happened while offline
2024-03-04 21:39:59.233 +05:00 [INF] Connected to WebSocket of NBXplorer (BTC)
2024-03-04 21:40:30.077 +05:00 [INF] New admin admin@admin.com registered
2024-03-04 21:45:57.893 +05:00 [INF] New user xx2@icloud.com invited by admin@admin.com
2024-03-04 21:48:45.891 +05:00 [INF] User a2a15fcee373-4b95-a145-aafd7107eb5b logged in
2024-03-04 21:50:06.774 +05:00 [INF] New user xxx3@btcpayserver.org invited by admin@admin.com
2024-03-04 21:51:31.432 +05:00 [INF] User logged out
2024-03-04 21:51:33.317 +05:00 [INF] User 282a4579-02d9-4b04-8ea3-ddd56188b344 logged in
```

**Comment by dstrukt** [comment]
> 1. Store owner doesn't get any notification or email if an invited successfully registered.

Can confirm, great catch!

> 2. Seems like permissions are not working correctly, I've invited user as a manager, but he only had guest access and no ability to view store settings.

I couldn't repro this one.

**Comment by dennisreimann** [comment]
> Store owner doesn't get any notification or email if an invited successfully registered.

Now they do and I've included managers as well, so owners and managers will get notifications and emails about a new user accepting the invite to a store. Done in b57336f.

grafik

> Seems like permissions are not working correctly, I've invited user as a manager, but he only had guest access and no ability to view store settings.

Yes, in the UI we mostly had `CanModifyStoreSettings` as a policy requirement. I've changed this in fa00c58 to make the read views available with the `CanViewStoreSettings` permission. This is a rather big change and will most likely interfere with the refactoring in #5782 — I think it will be best to first merge @kukks' PR and then rebase this one.

Nevertheless the general functionality can be tested already as both problems have been addressed.

> There are some things in the logs, but unsure what they're saying and if they're PR related:

These look like some kind of more general connection problem, which is most likely unrelated to the changes here.

**Comment by pavlenex** [comment]
Yeah under no circumstances should such low-permission role such as guest have insight into such important info like store settings.

**Comment by dennisreimann** [comment]
Their only practical permission would then be `canmodifyinvoices` and I wonder if we should simply remove the `Guest` role then and migrate them to `Employee`, so that we clean up everything as well.

**Comment by pavlenex** [comment]
@dennisreimann I'm afraid I don't have an answer to that. It does sound tempting to clean up those roles, but the `guest` role has been in there forever, and afaik there may be people building with API utilizing it. I'm unsure what sort of complexities will be involved to clean it up without breaking things up?

**Review by Kukks — APPROVED** [review]
i didn't see anything bad, although some preferences of mine don't match but that's fine (like the prefix subject stuff)

File: BTCPayServer/Controllers/UIStoresController.LightningLike.cs [hunk 1]
```diff
@@ -15,15 +16,16 @@
...
+using Microsoft.AspNetCore.Authorization;
 using Microsoft.AspNetCore.Mvc;
 using Microsoft.Extensions.DependencyInjection;
 
 namespace BTCPayServer.Controllers
 {
     public partial class UIStoresController
     {
-
         [HttpGet("{storeId}/lightning/{cryptoCode}")]
```
[the remark this answers is not in the document]
Comment (dennisreimann) [reply BTCPayServer/Controllers/UIStoresController.LightningLike.cs]: Agreed, but let's have a separate refactoring PR for that.

**Comment by dennisreimann** [comment]
@pavlenex Thanks to @kukks we now have readonly form-versions.

I've also added a better way of changing a user's role and tests for that. Can be user-tested again.

**Pull request merged**